File: Semaforo.java
import java.time.Instant;

public class Semaforo {
    private int cycleSeconds = 60;
    private int yellow = 3;
    private int allRed = 1;
    private int minGreen = 10;
    private int maxGreen = 40;

    private int greenNS;
    private int greenEW;
    private float loadNS;
    private float loadEW;
    private Instant lastUpdated;

    public int getCycleSeconds() { return cycleSeconds; }
    public void setCycleSeconds(int v) { cycleSeconds = Math.max(20, v); }
    public int getYellow() { return yellow; }
    public void setYellow(int v) { yellow = Math.max(2, v); }
    public int getAllRed() { return allRed; }
    public void setAllRed(int v) { allRed = Math.max(0, v); }
    public int getMinGreen() { return minGreen; }
    public void setMinGreen(int v) { minGreen = Math.max(5, v); }
    public int getMaxGreen() { return maxGreen; }
    public void setMaxGreen(int v) { maxGreen = Math.max(minGreen, v); }

    public int getGreenNS() { return greenNS; }
    public int getGreenEW() { return greenEW; }
    public float getLoadNS() { return loadNS; }
    public float getLoadEW() { return loadEW; }
    public Instant getLastUpdated() { return lastUpdated; }

    public void actualizarConCargas(float cargaNS, float cargaEW) {
        this.loadNS = clamp01(cargaNS);
        this.loadEW = clamp01(cargaEW);

        int budget = cycleSeconds - (2 * yellow) - allRed;
        if (budget < 2 * minGreen) {
            int needed = 2 * minGreen + allRed + 2 * yellow;
            cycleSeconds = Math.max(cycleSeconds, needed);
            budget = cycleSeconds - (2 * yellow) - allRed;
        }

        float dNS = 0.001f + loadNS;
        float dEW = 0.001f + loadEW;
        float sum = dNS + dEW;

        int gNS = Math.round(budget * (dNS / sum));
        int gEW = budget - gNS;

        gNS = clamp(gNS, minGreen, maxGreen);
        gEW = clamp(gEW, minGreen, maxGreen);

        int used = gNS + gEW;
        int delta = budget - used;
        if (delta != 0) {
            if (delta > 0) {
                if (loadNS >= loadEW && gNS < maxGreen) gNS += Math.min(delta, maxGreen - gNS);
                else if (gEW < maxGreen) gEW += Math.min(delta, maxGreen - gEW);
            } else {
                int take = -delta;
                if (loadNS <= loadEW && gNS > minGreen) gNS -= Math.min(take, gNS - minGreen);
                else if (gEW > minGreen) gEW -= Math.min(take, gEW - minGreen);
            }
        }

        this.greenNS = gNS;
        this.greenEW = gEW;
        this.lastUpdated = Instant.now();
    }

    private static float clamp01(float v) { return Math.max(0f, Math.min(1f, v)); }
    private static int clamp(int v, int lo, int hi) { return Math.max(lo, Math.min(hi, v)); }
}

File: Semaforo.js
// Semaforo.js
class Semaforo {
    constructor() {
        this.cycleSeconds = 60;
        this.yellow = 3;
        this.allRed = 1;
        this.minGreen = 10;
        this.maxGreen = 40;

        this._greenNS = 0; // Uso de _ para indicar que es 'privada'
        this._greenEW = 0;
        this._loadNS = 0.0;
        this._loadEW = 0.0;
        this._lastUpdated = null;
    }

    // Métodos Getters
    getCycleSeconds() { return this.cycleSeconds; }
    getYellow() { return this.yellow; }
    getAllRed() { return this.allRed; }
    getMinGreen() { return this.minGreen; }
    getMaxGreen() { return this.maxGreen; }
    getGreenNS() { return this._greenNS; }
    getGreenEW() { return this._greenEW; }
    getLoadNS() { return this._loadNS; }
    getLoadEW() { return this._loadEW; }
    getLastUpdated() { return this._lastUpdated; }

    // Métodos Setters
    setCycleSeconds(v) { this.cycleSeconds = Math.max(20, v); }
    setYellow(v) { this.yellow = Math.max(2, v); }
    setAllRed(v) { this.allRed = Math.max(0, v); }
    setMinGreen(v) { this.minGreen = Math.max(5, v); }
    setMaxGreen(v) { this.maxGreen = Math.max(this.minGreen, v); }

    /**
     * Limita un valor al rango [0, 1].
     * @param {number} v
     * @returns {number}
     */
    static _clamp01(v) { return Math.max(0.0, Math.min(1.0, v)); }

    /**
     * Limita un valor al rango [lo, hi].
     * @param {number} v
     * @param {number} lo
     * @param {number} hi
     * @returns {number}
     */
    static _clamp(v, lo, hi) { return Math.max(lo, Math.min(hi, v)); }

    /**
     * Actualiza los tiempos de luz verde basándose en la carga de tráfico.
     * @param {number} cargaNS Carga de tráfico Norte-Sur (0 a 1)
     * @param {number} cargaEW Carga de tráfico Este-Oeste (0 a 1)
     */
    actualizarConCargas(cargaNS, cargaEW) {
        this._loadNS = Semaforo._clamp01(cargaNS);
        this._loadEW = Semaforo._clamp01(cargaEW);

        let budget = this.cycleSeconds - (2 * this.yellow) - this.allRed;

        // Ajuste del ciclo si el presupuesto es insuficiente para el mínimo verde
        if (budget < 2 * this.minGreen) {
            const needed = 2 * this.minGreen + this.allRed + 2 * this.yellow;
            this.cycleSeconds = Math.max(this.cycleSeconds, needed);
            budget = this.cycleSeconds - (2 * this.yellow) - this.allRed;
        }

        // Cálculo de tiempos basado en la proporción de carga
        const dNS = 0.001 + this._loadNS;
        const dEW = 0.001 + this._loadEW;
        const sum = dNS + dEW;

        let gNS = Math.round(budget * (dNS / sum));
        let gEW = budget - gNS;

        // Aplicar límites MinGreen y MaxGreen
        gNS = Semaforo._clamp(gNS, this.minGreen, this.maxGreen);
        gEW = Semaforo._clamp(gEW, this.minGreen, this.maxGreen);

        // Redistribución del tiempo restante (delta)
        const used = gNS + gEW;
        let delta = budget - used;

        if (delta !== 0) {
            if (delta > 0) {
                // Hay tiempo extra: distribuirlo al carril con más carga si no ha alcanzado el máximo
                if (this._loadNS >= this._loadEW && gNS < this.maxGreen) {
                    gNS += Math.min(delta, this.maxGreen - gNS);
                } else if (gEW < this.maxGreen) {
                    gEW += Math.min(delta, this.maxGreen - gEW);
                }
            } else {
                // Hay exceso de tiempo asignado: restarlo al carril con menos carga si está por encima del mínimo
                const take = -delta;
                if (this._loadNS <= this._loadEW && gNS > this.minGreen) {
                    gNS -= Math.min(take, gNS - this.minGreen);
                } else if (gEW > this.minGreen) {
                    gEW -= Math.min(take, gEW - this.minGreen);
                }
            }
        }

        this._greenNS = gNS;
        this._greenEW = gEW;
        this._lastUpdated = Date.now(); // En JS, se usa Date.now() o new Date()
    }
}